File: Design-Patterns-Using-Go/Flyweight/FlyweightPatternRectangle.go
package main

import (
	"fmt"
	"math/rand"
)

// Shape interface
type Shape interface {
	Draw(x1, y1, x2, y2 int)
}

// Rectangle type
type Rectangle struct {
	colour string // Intrinsic State
}

func NewRectangle(colour string) *Rectangle {
	return &Rectangle{colour: colour}
}

func (r *Rectangle) Draw(x1, y1, x2, y2 int) {
	fmt.Printf("Draw rectangle colour:%s topleft: (%d,%d) rightBottom: (%d,%d)\n", r.colour, x1, y1, x2, y2)
}

// RectangleFactory type
type RectangleFactory struct {
	shapes map[string]Shape
}

func NewRectangleFactory() *RectangleFactory {
	return &RectangleFactory{
		shapes: make(map[string]Shape),
	}
}

func (rf *RectangleFactory) GetRectangle(colour string) Shape {
	if shape, exists := rf.shapes[colour]; exists {
		return shape
	}

	shape := NewRectangle(colour)
	rf.shapes[colour] = shape
	return shape
}

func (rf *RectangleFactory) GetCount() int {
	return len(rf.shapes)
}

// Client Code
func main() {
	factory := NewRectangleFactory()
	for i := 0; i < 1000; i++ {
		colour := fmt.Sprintf("%d", rand.Intn(1000))
		rect := factory.GetRectangle(colour)
		rect.Draw(rand.Intn(100), rand.Intn(100), rand.Intn(100), rand.Intn(100))
	}
	fmt.Println(factory.GetCount())
}

File: Design-Patterns-Using-Cpp/Flyweight/FlyweightPatternRectangle.cpp
#include <iostream>
#include <map>
#include <random>

// Shape abstract class
class Shape {
protected:
    int colour; // Intrinsic State

public:
    Shape(int colour) : colour(colour) {}

    virtual void draw(int x1, int y1, int x2, int y2) = 0; // Extrinsic State
};

// Rectangle class
class Rectangle : public Shape {
public:
    Rectangle(int colour) : Shape(colour) {}

    void draw(int x1, int y1, int x2, int y2) override {
        std::cout << "Draw Rectangle colour:" << colour << " topleft: (" << x1 << "," << y1 << ") rightBottom: (" << x2 << "," << y2 << ")\n";
    }
};

// RectangleFactory class
class RectangleFactory {
private:
    std::map<int, Shape*> shapes;

public:
    Shape* getRectangle(int colour) {
        if (shapes.find(colour) == shapes.end()) {
            shapes[colour] = new Rectangle(colour);
        }
        return shapes[colour];
    }

    int getCount() const {
        return shapes.size();
    }

    ~RectangleFactory() {
        for (auto& pair : shapes) {
            delete pair.second;
        }
    }
};

// Client code
int main() {
    RectangleFactory factory;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> dis(0, 999);

    for (int i = 0; i < 1000; ++i) {
        Shape* rect = factory.getRectangle(dis(gen));
        rect->draw(dis(gen), dis(gen), dis(gen), dis(gen));
    }

    std::cout << factory.getCount() << std::endl;

    return 0;
}
